presto-native-execution/presto_cpp/main/operators/PartitionAndSerialize.cpp

```diff
@@ -71,44 +71,99 @@ class PartitionAndSerializeOperator : public Operator {
+  /// serialize input elements only.
+  ///
+  /// Branch 2) 'replicateNullsAndAny_' is true and output vector
+  /// is of reasonable size compared to desired output size per batch (see:
+  /// Operator::outputBatchRows).
+  ///
+  /// Branch 3) 'replicateNullsAndAny_' is true and
+  /// output vector will be significantly large, Do this in steps:
+  ///
+  /// Step 1) Same as Branch 1.
+  ///
+  /// Step 2) Replicate null values for 'nextPartitionId_' and repeat with
+  /// nextPartitionId_ + 1
   RowVectorPtr getOutput() override {
     if (!input_) {
       return nullptr;
     }
-    collectNullRows();
 
-    const vector_size_t maxOutputSize = calculateMaxOutputSize();
+    vector_size_t numInput = input_->size();
+    velox::row::UnsafeRowFast unsafeRow(reorderInputsIfNeeded());
+    calculateSerializedRowSize(unsafeRow);
 
-    // TODO Reuse output vector.
-    auto output =
-        BaseVector::create<RowVector>(outputType_, maxOutputSize, pool());
+    std::optional<uint64_t> averageRowSize;
+    if (!cumulativeRowSizes_.empty()) {
+      averageRowSize = cumulativeRowSizes_[cumulativeRowSizes_.size() - 1] /
+          cumulativeRowSizes_.size();
+    }
+    const vector_size_t maxOutputSize = outputBatchRows(averageRowSize);
+    auto output = createOutputRowVector(maxOutputSize);
     auto partitionVector = output->childAt(0)->asFlatVector<int32_t>();
-    partitionVector->resize(maxOutputSize);
     auto dataVector = output->childAt(1)->asFlatVector<StringView>();
-    dataVector->resize(maxOutputSize);
-
-    computePartitions(*partitionVector);
-    serializeRows(*dataVector);
-
-    if (replicateNullsAndAny_ && numPartitions_ > 1) {
-      auto numOutputRows =
-          replicateRowsWithNullPartitionKeys(*dataVector, *partitionVector);
-      // If numOutputRows equal to the input row count, then
-      // 'replicateRowsWithNullPartitionKeys' did not insert any row into the
-      // output vector, thus we need to call into
-      // 'replicateAnyToEmptyPartitions'
-      if (numOutputRows == input_->size()) {
-        numOutputRows = replicateAnyToEmptyPartitions(
-            *partitionVector, *dataVector, numOutputRows);
+
+    if (processingFirstBatch_) {
+      computePartitions(*partitionVector);
+      serializeRows(unsafeRow, *dataVector);
+      processingFirstBatch_ = false;
+      // We can fit remaining elements into the output vector in the same pass.
+      if (replicateNullsAndAny_ && maxOutputSize > numInput) {
+        auto numOutputRows = replicateRowsWithNullPartitionKeys(
+            *partitionVector, *dataVector, 0, numPartitions_, numInput);
+        // If numOutputRows equal to the input row count, then
+        // 'replicateRowsWithNullPartitionKeys' did not insert any row into
+        // the output vector, thus we need to call into
+        // 'replicateAnyToEmptyPartitions'
+        if (numOutputRows == input_->size()) {
+          numOutputRows = replicateAnyToEmptyPartitions(
```

Comment:
How do we ensure that we replicate 'any' row for each partition once? Where do we track the fact that we already replicated 'any' for a partition (possibly when processing previous batch of input)?